File: IEEEXtreme14_15/UoWM-Kastoria/Rigged-Dice/rigged_dice.java
//
//   CSAcademy Problem - Rigged Dice
//   From: IEEEXtreme 15.0
//   Solved By: Dimitrios and Petros Papakonstantinou
//   Submission Score: 100/100 (1183ms - 36.1MB)
//

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);

    // Read the number of games
    int game_num = scanner.nextInt();
    scanner.nextLine(); // Consume the newline after the number input

    // Process each game
    for (int game = 0; game < game_num; ++game) {
      // Read the number of rounds for this game
      int n = scanner.nextInt();
      scanner.nextLine(); // Consume the newline after the number input

      // Lists to store dice rolls for Alice and Bob
      List<Integer> d1 =
          new ArrayList<>(); // Represents the rolls for Alice's current die
      List<Integer> d2 =
          new ArrayList<>(); // Represents the rolls for Bob's current die

      int alice_total = 0,
          bob_total = 0; // Keep track of total scores for Alice and Bob
      boolean current_alice_die1 =
          true; // Flag to determine whose roll is stored in d1 and d2

      // Read and process each round
      for (int i = 0; i < n; ++i) {
        // Read the line of input containing Alice's and Bob's rolls
        String line = scanner.nextLine();
        String[] tokens = line.split(
            " "); // Split the input line into Alice's and Bob's rolls

        // Parse the rolls from the input
        int alice = Integer.parseInt(tokens[0]);
        int bob = Integer.parseInt(tokens[1]);

        // Update total scores for Alice and Bob
        alice_total += alice;
        bob_total += bob;

        // Assign the current round's rolls to the respective lists based on
        // whose die is active
        if (current_alice_die1) {
          d1.add(alice); // Alice's roll goes to d1, Bob's to d2
          d2.add(bob);
        } else {
          d2.add(alice); // Switch roles: Alice's roll to d2, Bob's to d1
          d1.add(bob);
        }

        // If Alice and Bob's total scores are unequal, switch the dice
        // assignment for the next round
        if (alice_total != bob_total) {
          current_alice_die1 =
              !current_alice_die1; // Toggle dice assignment for the next round
        }
      }

      // Count the number of sixes in Alice's die (d1) and Bob's die (d2)
      long d1_count =
          d1.stream().filter(x -> x == 6).count(); // Count sixes in d1
      long d2_count =
          d2.stream().filter(x -> x == 6).count(); // Count sixes in d2

      // Determine the winner based on which die (d1 or d2) has more sixes
      if (d1_count > d2_count) {
        System.out.println("1"); // Alice wins if d1 has more sixes
      } else {
        System.out.println(
            "2"); // Bob wins if d2 has more sixes, or if they are equal
      }
    }

    scanner.close(); // Close the scanner after input is processed
  }
}

File: IEEEXtreme14_15/UoWM-Kastoria/Rigged-Dice/rigged_dice.js
//
//   CSAcademy Problem - Rigged Dice
//   From: IEEEXtreme 15.0
//   Solved By: Dimitrios and Petros Papakonstantinou
//   Submission Score: 100/100 (1957ms - 36.6MB)
//

const readline = require('readline');

// Create an interface to read from standard input and output
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let gameNum;           // Total number of games to be played
let currentGame = 0;    // Index of the current game being processed
let n;                 // Number of rounds in the current game
let currentRoll = 0;    // Counter for the number of rounds processed in the current game

// Function to process a single game
function processGame() {
  let d1 = [], d2 = [];               // Arrays to store dice rolls for Alice and Bob
  let aliceTotal = 0, bobTotal = 0;    // Variables to track total scores for Alice and Bob
  let currentAliceDie1 = true;         // Flag to toggle whose roll is added to d1/d2

  // Function to process a single round of rolls (for Alice and Bob)
  function processRoll() {
    rl.question('', (line) => {
      // Parse the current round's dice rolls for Alice and Bob
      const [alice, bob] = line.split(' ').map(Number);

      // Update the total scores for Alice and Bob
      aliceTotal += alice;
      bobTotal += bob;

      // Depending on the value of currentAliceDie1, store Alice's and Bob's rolls in d1 or d2
      if (currentAliceDie1) {
        d1.push(alice);  // Add Alice's roll to d1 and Bob's to d2
        d2.push(bob);
      } else {
        d2.push(alice);  // Add Alice's roll to d2 and Bob's to d1
        d1.push(bob);
      }

      // If the total scores for Alice and Bob are unequal, switch the die assignments for the next round
      if (aliceTotal !== bobTotal) {
        currentAliceDie1 = !currentAliceDie1;  // Toggle die assignment
      }

      currentRoll++;  // Increment the round counter

      // If there are more rounds left in the current game, process the next roll
      if (currentRoll < n) {
        processRoll();
      } else {
        // If all rounds for the current game are processed, count the number of sixes in each die
        const d1Count = d1.filter(x => x === 6).length;  // Count the sixes in d1
        const d2Count = d2.filter(x => x === 6).length;  // Count the sixes in d2

        // Determine the winner based on the count of sixes
        console.log(d1Count > d2Count ? '1' : '2');  // Alice wins if d1 has more sixes, otherwise Bob wins

        currentGame++;  // Move to the next game

        // If more games are left, start the next game
        if (currentGame < gameNum) {
          startNextGame();
        } else {
          rl.close();  // Close the readline interface if all games are processed
        }
      }
    });
  }

  // Read the number of rounds (n) for the current game and start processing rolls
  rl.question('', (input) => {
    n = parseInt(input);  // Parse the number of rounds
    currentRoll = 0;      // Reset the round counter for the new game
    processRoll();        // Start processing the first roll
  });
}

// Function to start the next game
function startNextGame() {
  processGame();  // Process a new game
}

// Start by reading the number of games
rl.question('', (input) => {
  gameNum = parseInt(input);  // Parse the number of games
  startNextGame();            // Start processing the first game
});
